lib/Parse/ParseDecl.cpp

```diff
@@ -437,6 +437,54 @@ bool Parser::parseSpecializeAttribute(swift::tok ClosingBrace, SourceLoc AtLoc,
   return true;
 }
 
+bool Parser::parseImplementsAttribute(SourceLoc AtLoc, SourceLoc Loc,
+                                      ImplementsAttr *&Attr) {
+  StringRef AttrName = "_implements";
+  Attr = nullptr;
+
+  if (Tok.isNot(tok::l_paren)) {
+    diagnose(Loc, diag::attr_expected_lparen, AttrName,
+             /*DeclModifier=*/false);
+    return false;
+  }
+
+  SourceLoc lParenLoc = consumeToken();
+
+  if (!Tok.is(tok::string_literal)) {
+    diagnose(Loc, diag::attr_expected_string_literal, AttrName);
+    return false;
+  }
+
+  Optional<StringRef> ImplName =
+    getStringLiteralIfNotInterpolated(*this, Loc, Tok, AttrName);
+
+  // getStringLiteralIfNotInterpolated already emits a diagnostic.
+  if (!ImplName.hasValue())
+    return false;
+
+  consumeToken(tok::string_literal);
+
+  SourceLoc rParenLoc;
+  if (!consumeIf(tok::r_paren, rParenLoc)) {
+    diagnose(lParenLoc, diag::attr_expected_rparen, AttrName,
+             /*DeclModifier=*/false);
+    return false;
+  }
+
+  // We parse a ParsedDeclName here just to make sure it's well formed early; we
+  // store it in the attr as a StringRef and re-parse on use, so as not to leak
+  // the ParsedDeclName type into Attr.h
+  ParsedDeclName parsedName = parseDeclName(ImplName.getValue());
```

Comment:
Perhaps the semantic checking can be done in lib/Sema/TypeCheckAttr.cpp, so you're not simply throwing away the result.